gcp/warehouse.go

```diff
@@ -0,0 +1,158 @@
+// Copyright 2018 Google LLC
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     https://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+// Package gcp abstracts interacting with certain aspects of Google Cloud
+// Platform, such as creating service account keys and access tokens.
+package gcp
+
+import (
+	"context"
+	"encoding/base64"
+	"encoding/hex"
+	"fmt"
+	"net/http"
+	"path"
+
+	"golang.org/x/crypto/sha3"
+	cloudresourcemanager "google.golang.org/api/cloudresourcemanager/v1"
+	"google.golang.org/api/googleapi"
+	iam "google.golang.org/api/iam/v1"
+	iamcredentials "google.golang.org/api/iamcredentials/v1"
+)
+
+// WarehouseOptions is used with NewWarehouse to configure a new warehouse.
+type WarehouseOptions struct {
+	Project     string
+	DefaultRole string
+	Scopes      []string
+}
+
+// Warehouse is used to create Google Cloud Platform Service Account keys and
+// access tokens associated with a specific identity.
+type Warehouse struct {
+	opts  WarehouseOptions
+	iam   *iam.Service
+	creds *iamcredentials.Service
+	crm   *cloudresourcemanager.Service
+}
+
+// NewWarehouse creates a new Warehouse using the provided client and options.
+func NewWarehouse(client *http.Client, opts *WarehouseOptions) (*Warehouse, error) {
+	iamSvc, err := iam.New(client)
+	if err != nil {
+		return nil, fmt.Errorf("creating IAM client: %v", err)
+	}
+
+	creds, err := iamcredentials.New(client)
+	if err != nil {
+		return nil, fmt.Errorf("creating IAM credentials client: %v", err)
+	}
+
+	crm, err := cloudresourcemanager.New(client)
+	if err != nil {
+		return nil, fmt.Errorf("creating cloud resource manager client: %v", err)
+	}
+
+	return &Warehouse{
+		opts:  *opts,
+		iam:   iamSvc,
+		creds: creds,
+		crm:   crm,
+	}, nil
+}
+
+// GetAccountKey returns a service account key associated with id.
+func (wh *Warehouse) GetAccountKey(ctx context.Context, id string) ([]byte, error) {
+	account, err := wh.getBackingAccount(ctx, id)
+	if err != nil {
+		return nil, fmt.Errorf("getting backing account: %v", err)
+	}
+
+	// TODO: Remove old keys from service account before creating new keys.
```

Comment:
Use TODO(#X) style and file issues

What happens as it stands?  Do the old keys get disabled automatically or new key creation fails?